refactor(libsinsp): manage memory in ast/parser via unique_ptrs

LucaGuerra opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

/kind bug

/kind cleanup

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

/area libsinsp
/area tests

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**Does this PR require a change in the driver versions?**

<!--
Please remove the leading whitespace before the `/version <>` you uncommented.
-->

**What this PR does / why we need it**:

This refactor is also the larger part of the effort to make our tests pass under AddressSanitizer.

All the `ast::expr` classes were holding pointers or vectors of pointers allocated manually via `new` and removed via `delete`, sometimes through virtual destructors. This should be fine in the happy path where everything works as planned, but becomes an issue where any unexpected behavior happens, such as any `throw` in the parsers. In that case, any allocated memory would not be freed and thus leaked. The issue is apparent in tests that exercise parsing errors or other edge cases when using LeakSanitizer which will loudly complain about a lot of leaked memory. This PR aims to remove any manual memory management in those classes and replace it with `std::unique_ptr`s. 

I decided to solve the issue this way because any expression node owns the nodes below it, and the parser is recursive, normally operating by allocating the lower nodes and then assigning them to the nodes one level up. This also has the added benefit of removing virtual destructors since we don't need them anymore.

The contract with functions that return or get parameters as `std::unique_ptr` works as follows:
* If the function/constructor _requires_ a `std::unique_ptr` it means that it's going to acquire full ownership of that pointer, which of course needs to be `std::move`d there.
* If the function _returns_ a `std::unique_ptr` it means that it will give ownership of the object to the caller
* Same goes for `std::vector<std::unique_ptr<>>`

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

See inline comments.

**Does this PR introduce a user-facing change?**:

<!--
If no, you have to do nothing.
If yes, a release note is required:
Delete `NONE` and enter your extended release note in the block below.
Please note, the release note follows the "conventional commit specification" (https://www.conventionalcommits.org/en/v1.0.0/):
For example: `fix: broken link`.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
NONE
```

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/535#" title="Author self-approved">LucaGuerra</a>*
**Once this PR has been reviewed and has the lgtm label**, please assign andreagit97 for approval by writing `/assign @andreagit97` in a comment. For more information see:[The Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["andreagit97"]} -->

LucaGuerra commented on userspace/libsinsp/filter/ast.h:
```diff
@@ -181,24 +187,40 @@ struct SINSP_PUBLIC or_expr: expr
-        return o != nullptr && std::equal(
-            children.begin(), children.end(), 
-            o->children.begin(), compare);
+        if (o == nullptr)
+        {
+            return false;
+        }
+
+        for (size_t i = 0; i < children.size(); i++)
+        {
+            if (o->children.size() <= i)
+            {
+                return false;
+            }
+
+            if (!compare(children[i].get(), o->children[i].get()))
+            {
+                return false;
+            }
+        }
+
+        return true;
```
This is not as clever and elegant as before, but I think it makes sense in this case to be completely explicit about the pointer usage.

LucaGuerra commented on userspace/libsinsp/test/filter_parser.ut.cpp:
```diff
@@ -385,57 +383,66 @@ TEST(parser, parse_position_info)
 // complex test case with all supported node types
 TEST(parser, expr_all_node_types)
 {
+	std::vector<std::unique_ptr<expr>> and_children;
+	and_children.push_back(unary_check_expr::create("evt.name", "", "exists"));
+	and_children.push_back(binary_check_expr::create("evt.type", "", "in", list_expr::create({"a", "b"})));
+	and_children.push_back(not_expr::create(binary_check_expr::create("evt.dir", "", "=", value_expr::create("<"))));
+
+	std::vector<std::unique_ptr<expr>> or_children;
+	or_children.push_back(and_expr::create(and_children));
+	or_children.push_back(binary_check_expr::create("proc.name", "", "=", value_expr::create("cat")));
+
+	std::unique_ptr<expr> ast = or_expr::create(or_children);
```
This is also less readable, but the simplest way I found to pass a `std::vector<std::unique_ptr>>` around to be moved seems to be via lvalue reference, so it needs a named object

LucaGuerra commented on userspace/libsinsp/filter/ast.cpp:
```diff
@@ -180,64 +180,62 @@ std::string libsinsp::filter::ast::as_string(ast::expr &e)
         }
     } visitor;
 
-    visitor.m_last_node = NULL;
```
note for the reviewer: this looked redundant, especially because a `unique_ptr` starts null/without any object, but please double check

FedeDP commented:
/cc @jasondellaluce

jasondellaluce reviewed (commented):
This looks great to me! I left you a minor suggestion but this is very helpful.

Side note: it would probably require few refactors in Falco too, but not a big deal.

LucaGuerra commented on userspace/libsinsp/filter/ast.h:
```diff
@@ -151,27 +144,40 @@ struct SINSP_PUBLIC and_expr: expr
+
+        for (size_t i = 0; i < children.size(); i++)
+        {
+            if (o->children.size() <= i)
```
definitely!

poiana commented:
LGTM label has been added.  <details>Git tree hash: a7a91d6c927305c33a0e98e3b04d2bedc4c04651</details>

poiana commented:
New changes are detected. LGTM label has been removed.

LucaGuerra commented:
Addressed comment, plus another small thing that surfaced from myself getting confused by my own refactoring: added an explicit `const` when using a `std::vector<std::unique_ptr>` in a read only way so it is clear that the specific function is not transferring ownership (it could if it was non-`const`)

poiana commented:
LGTM label has been added.  <details>Git tree hash: 247ba8d0cded094647f5682458e039156dc52ca7</details>

leogr reviewed (approved):
LGTM

FedeDP reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/535#pullrequestreview-1080069863" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/535#" title="Author self-approved">LucaGuerra</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [FedeDP,LucaGuerra]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
